File: DefReader.cpp
#include "DefReader.hpp"
#include "LLParser.hpp"
#include "ExtendedGrammar.hpp"

#include <fstream>
#include <vector>
#include <iostream>

std::string escape(const std::string &input)
{
    std::string result;

    for(char c : input) {
        switch(c) {
            case ' ': result.append("\\s"); break;
            case '+':
            case '*':
            case '?':
            case '(':
            case ')':
            case '[':
            case ']': result.push_back('\\');
            default: result.push_back(c); break;
        }
    }

    return result;
}

struct StringData
{
    StringData(const std::string &t) : text(t) {}

    std::string text;
};

struct DefNode {
    enum class Type {
        List,
        Terminal,
        Nonterminal,
        Literal,
        Regex,
        Pattern,
        Rule,
        RhsSequence,
        RhsOneOf,
        RhsZeroOrMore,
        RhsOneOrMore,
        RhsZeroOrOne
    };

    template<typename ...Children> DefNode(Type t, Children&&... c) : type(t) {
        children.reserve(sizeof...(c));
        (children.push_back(std::move(c)), ...);
    }
    DefNode(Type t, const std::string &s) : type(t), string(s) {}

    Type type;
    std::vector<std::unique_ptr<DefNode>> children;
    std::string string;
};

std::unique_ptr<ExtendedGrammar::RhsNode> ZeroOrMore(std::unique_ptr<ExtendedGrammar::RhsNode> &&node) {
    return std::make_unique<ExtendedGrammar::RhsNodeChild>(ExtendedGrammar::RhsNode::Type::ZeroOrMore, std::move(node));
};

std::unique_ptr<ExtendedGrammar::RhsNode> OneOrMore(std::unique_ptr<ExtendedGrammar::RhsNode> &&node) {
    return std::make_unique<ExtendedGrammar::RhsNodeChild>(ExtendedGrammar::RhsNode::Type::OneOrMore, std::move(node));
};

std::unique_ptr<ExtendedGrammar::RhsNode> ZeroOrOne(std::unique_ptr<ExtendedGrammar::RhsNode> &&node) {
    return std::make_unique<ExtendedGrammar::RhsNodeChild>(ExtendedGrammar::RhsNode::Type::ZeroOrOne, std::move(node));
};

template<typename...Args> std::unique_ptr<ExtendedGrammar::RhsNode> OneOf(Args&&... nodes) {
    return std::make_unique<ExtendedGrammar::RhsNodeChildren>(ExtendedGrammar::RhsNode::Type::OneOf, std::forward<Args>(nodes)...);
};

template<typename...Args> std::unique_ptr<ExtendedGrammar::RhsNode> Sequence(Args&&... nodes) {
    return std::make_unique<ExtendedGrammar::RhsNodeChildren>(ExtendedGrammar::RhsNode::Type::Sequence, std::forward<Args>(nodes)...);
};

std::unique_ptr<ExtendedGrammar::RhsNode> createRhsNode(
    const DefNode &defNode,
    std::map<std::string, unsigned int> &ruleMap,
    std::map<std::string, unsigned int> &terminalMap,
    std::map<std::string, unsigned int> &anonymousTerminalMap,
    std::vector<std::string> &terminals,
    std::vector<std::string> &terminalNames)
{
    switch(defNode.type) {
        case DefNode::Type::Terminal:
        {
            unsigned int index = UINT_MAX;
            auto it = terminalMap.find(defNode.string);
            if(it != terminalMap.end()) {
                index = it->second;
            }
            return std::make_unique<ExtendedGrammar::RhsNodeSymbol>(ExtendedGrammar::RhsNodeSymbol::SymbolType::Terminal, index);
        }

        case DefNode::Type::Nonterminal:
        {
            unsigned int index = UINT_MAX;
            auto it = ruleMap.find(defNode.string);
            if(it != ruleMap.end()) {
                index = it->second;
            }
            return std::make_unique<ExtendedGrammar::RhsNodeSymbol>(ExtendedGrammar::RhsNodeSymbol::SymbolType::Nonterminal, index);
        }

        case DefNode::Type::Literal:
        {
            std::string text = escape(defNode.string);
            unsigned int index;

            auto it = anonymousTerminalMap.find(text);
            if(it == anonymousTerminalMap.end()) {
                terminals.push_back(text);
                terminalNames.push_back(text);
                index = anonymousTerminalMap[text] = (unsigned int)(terminals.size() - 1);
            } else {
                index = it->second;
            }
            return std::make_unique<ExtendedGrammar::RhsNodeSymbol>(ExtendedGrammar::RhsNodeSymbol::SymbolType::Terminal, index);
        }

        case DefNode::Type::RhsSequence:
        {
            std::unique_ptr<ExtendedGrammar::RhsNodeChildren> node = std::make_unique<ExtendedGrammar::RhsNodeChildren>(ExtendedGrammar::RhsNode::Type::Sequence);
            for(auto &child : defNode.children) {
                std::unique_ptr<ExtendedGrammar::RhsNode> childRhsNode = createRhsNode(*child, ruleMap, terminalMap, anonymousTerminalMap, terminals, terminalNames);
                node->children.push_back(std::move(childRhsNode));
            }
            return node;
        }

        case DefNode::Type::RhsOneOf:
        {
            std::unique_ptr<ExtendedGrammar::RhsNodeChildren> node = std::make_unique<ExtendedGrammar::RhsNodeChildren>(ExtendedGrammar::RhsNode::Type::OneOf);
            for(auto &child : defNode.children) {
                std::unique_ptr<ExtendedGrammar::RhsNode> childRhsNode = createRhsNode(*child, ruleMap, terminalMap, anonymousTerminalMap, terminals, terminalNames);
                node->children.push_back(std::move(childRhsNode));
            }
            return node;
        }

        case DefNode::Type::RhsZeroOrMore:
        {
            std::unique_ptr<ExtendedGrammar::RhsNode> child = createRhsNode(*defNode.children[0], ruleMap, terminalMap, anonymousTerminalMap, terminals, terminalNames);
            return std::make_unique<ExtendedGrammar::RhsNodeChild>(ExtendedGrammar::RhsNode::Type::ZeroOrMore, std::move(child));
        }

        case DefNode::Type::RhsOneOrMore:
        {
            std::unique_ptr<ExtendedGrammar::RhsNode> child = createRhsNode(*defNode.children[0], ruleMap, terminalMap, anonymousTerminalMap, terminals, terminalNames);
            return std::make_unique<ExtendedGrammar::RhsNodeChild>(ExtendedGrammar::RhsNode::Type::OneOrMore, std::move(child));
        }

        case DefNode::Type::RhsZeroOrOne:
        {
            std::unique_ptr<ExtendedGrammar::RhsNode> child = createRhsNode(*defNode.children[0], ruleMap, terminalMap, anonymousTerminalMap, terminals, terminalNames);
            return std::make_unique<ExtendedGrammar::RhsNodeChild>(ExtendedGrammar::RhsNode::Type::ZeroOrOne, std::move(child));
        }
    }
    return std::unique_ptr<ExtendedGrammar::RhsNode>();
};

DefReader::DefReader(const std::string &filename)
{
    std::vector<std::string> grammarTerminals{
        "epsilon",
        "terminal",
        "nonterminal",
        "colon",
        "pipe",
        "lparen",
        "rparen",
        "star",
        "plus",
        "question",
        "literal",
        "regex",
        "newline",
        "end"
    };

    auto tokenIndex = [&](const std::string &name) {
        for(unsigned int i=0; i<grammarTerminals.size(); i++) {
            if(grammarTerminals[i] == name) {
                return i;
            }
        }
        return Tokenizer::InvalidTokenValue;
    };

    auto pattern = [&](const std::string &regex, const std::string &name) {
        return Tokenizer::Pattern{regex, name, tokenIndex(name)};
    };

    std::vector<Tokenizer::Configuration> configurations{
        Tokenizer::Configuration{std::vector<Tokenizer::Pattern>{
            pattern("0", "epsilon"),
            pattern("\\w+", "terminal"),
            pattern("<\\w+>", "nonterminal"),
            pattern(":", "colon"),
            pattern("\\|", "pipe"),
            pattern("\\(", "lparen"),
            pattern("\\)", "rparen"),
            pattern("\\+", "plus"),
            pattern("\\*", "star"),
            pattern("\\?", "question"),
            pattern("'[^']+'", "literal"),
            pattern("\\s", "whitespace")  
        }},
        Tokenizer::Configuration{std::vector<Tokenizer::Pattern>{
            pattern("\\S+", "regex"),
            pattern("\\s", "whitespace")
        }}
    };

    Tokenizer tokenizer(std::move(configurations), tokenIndex("end"), tokenIndex("newline"));

    std::vector<ExtendedGrammar::Rule> grammarRules;
    grammarRules.push_back(ExtendedGrammar::Rule{"root"});
    grammarRules.push_back(ExtendedGrammar::Rule{"definitions"});
    grammarRules.push_back(ExtendedGrammar::Rule{"pattern"});
    grammarRules.push_back(ExtendedGrammar::Rule{"rule"});
    grammarRules.push_back(ExtendedGrammar::Rule{"rhs"});
    grammarRules.push_back(ExtendedGrammar::Rule{"rhsSuffix"});
    grammarRules.push_back(ExtendedGrammar::Rule{"rhsSymbol"});

    auto ruleIndex = [&](const std::string &name) {
        for(unsigned int i=0; i<grammarRules.size(); i++) {
            if(grammarRules[i].lhs == name) {
                return i;
            }
        }
        return UINT_MAX;
    };

    auto T = [&](const std::string &name) -> std::unique_ptr<ExtendedGrammar::RhsNode>  {
        return std::make_unique<ExtendedGrammar::RhsNodeSymbol>(ExtendedGrammar::RhsNodeSymbol::SymbolType::Terminal, tokenIndex(name));
    };

    auto N = [&](const std::string &name) -> std::unique_ptr<ExtendedGrammar::RhsNode>  {
        return std::make_unique<ExtendedGrammar::RhsNodeSymbol>(ExtendedGrammar::RhsNodeSymbol::SymbolType::Nonterminal, ruleIndex(name));
    };

    auto SetRule = [&](const std::string &name, std::unique_ptr<ExtendedGrammar::RhsNode> &&node) {
        grammarRules[ruleIndex(name)].rhs = std::move(node);
    };

    SetRule("root", Sequence(N("definitions"), T("end")));
    SetRule("definitions", ZeroOrMore(OneOf(N("pattern"), N("rule"), T("newline"))));
    SetRule("pattern", Sequence(T("terminal"), T("colon"), T("regex"), T("newline")));
    SetRule("rule", Sequence(T("nonterminal"), T("colon"), N("rhs"), T("newline")));
    SetRule("rhs", OneOrMore(N("rhsSuffix")));
    SetRule("rhsSuffix", 
        Sequence(
            N("rhsSymbol"),
            ZeroOrMore(
                OneOf(T("star"), T("plus"), T("question"))
            )
        )
    );
    SetRule("rhsSymbol",
        OneOf(
            T("terminal"),
            T("nonterminal"),
            T("literal"),
            Sequence(
                T("lparen"), N("rhs"), ZeroOrMore(Sequence(T("pipe"), N("rhs"))), T("rparen")
            )
        )
    );

    ExtendedGrammar extendedGrammar(std::move(grammarTerminals), std::move(grammarRules), 0);
    std::unique_ptr<Grammar> grammar = extendedGrammar.makeGrammar();
    LLParser parser(*grammar);
    
    std::ifstream file(filename);
    Tokenizer::Stream<StringData> stream(tokenizer, file);

    stream.addDecorator("terminal", 0, [](const std::string &text) {
        return std::make_unique<StringData>(text);
    });
    stream.addDecorator("nonterminal", 0, [](const std::string &text) {
        return std::make_unique<StringData>(text.substr(1, text.size() - 2));
    });
    stream.addDecorator("literal", 0, [](const std::string &text) {
        return std::make_unique<StringData>(text.substr(1, text.size() - 2));
    });
    stream.addDecorator("regex", 1, [](const std::string &text) {
        return std::make_unique<StringData>(text);
    });

    LLParser::ParseSession<DefNode, StringData> session(parser);
    session.addMatchListener("pattern", [&](unsigned int symbol) {
        if(symbol == 1) stream.setConfiguration(1);
        else if(symbol == 2) stream.setConfiguration(0);
    });

    session.addTerminalDecorator("terminal", [](const StringData &stringData) {
        return std::make_unique<DefNode>(DefNode::Type::Terminal, stringData.text);
    });
    session.addTerminalDecorator("nonterminal", [](const StringData &stringData) {
        return std::make_unique<DefNode>(DefNode::Type::Nonterminal, stringData.text);
    });
    session.addTerminalDecorator("literal", [](const StringData &stringData) {
        return std::make_unique<DefNode>(DefNode::Type::Literal, stringData.text);
    });
    session.addTerminalDecorator("regex", [](const StringData &stringData) {
        return std::make_unique<DefNode>(DefNode::Type::Regex, stringData.text);
    });

    session.addReducer("root", [](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        return std::move(items[0].data);
    });
    session.addReducer("definitions", [](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        std::unique_ptr<DefNode> node = std::make_unique<DefNode>(DefNode::Type::List);
        for(unsigned int i=0; i<numItems; i++) {
            if(items[i].data) {
                node->children.push_back(std::move(items[i].data));
            }
        }
        return node;
    });
    session.addReducer("pattern", [](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        return std::make_unique<DefNode>(DefNode::Type::Pattern, std::move(items[0].data), std::move(items[2].data));
    });
    session.addReducer("rule", [](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        return std::make_unique<DefNode>(DefNode::Type::Rule, std::move(items[0].data), std::move(items[2].data));
    });
    session.addReducer("rhs", [](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        if(numItems == 1) {
            return std::move(items[0].data);
        } else {
            std::unique_ptr<DefNode> node = std::make_unique<DefNode>(DefNode::Type::RhsSequence);
            for(unsigned int i=0; i<numItems; i++) {
                node->children.push_back(std::move(items[i].data));
            }
            return node;
        }
    });
    unsigned int star = grammar->terminalIndex("star");
    unsigned int plus = grammar->terminalIndex("plus");
    unsigned int question = grammar->terminalIndex("question");
    session.addReducer("rhsSuffix", [&](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        std::unique_ptr<DefNode> node = std::move(items[0].data);
        for(unsigned int i=1; i<numItems; i++) {
            if(items[i].index == star) {
                node = std::make_unique<DefNode>(DefNode::Type::RhsZeroOrMore, std::move(node));
            } else if(items[i].index == plus) {
                node = std::make_unique<DefNode>(DefNode::Type::RhsOneOrMore, std::move(node));
            } else if(items[i].index == question) {
                node = std::make_unique<DefNode>(DefNode::Type::RhsZeroOrOne, std::move(node));
            }
        }
        return node;
    });
    unsigned int lparen = grammar->terminalIndex("lparen");
    session.addReducer("rhsSymbol", [&](LLParser::ParseItem<DefNode> *items, unsigned int numItems) {
        if(numItems == 1) {
            return std::move(items[0].data);        
        } else if(numItems == 3) {
            return std::move(items[1].data);
        } else {
            std::unique_ptr<DefNode> node = std::make_unique<DefNode>(DefNode::Type::RhsOneOf);
            for(unsigned int i=1; i<numItems; i+=2) {
                node->children.push_back(std::move(items[i].data));
            }
            return node;
        }

        return std::unique_ptr<DefNode>();
    });

    std::unique_ptr<DefNode> node = session.parse(stream);

    std::map<std::string, unsigned int> terminalMap;
    std::map<std::string, unsigned int> anonymousTerminalMap;
    std::vector<std::string> terminals;
    std::vector<std::string> terminalNames;

    for(const auto &definition: node->children) {
        if(definition->type == DefNode::Type::Pattern) {
            terminals.push_back(definition->children[1]->string);
            terminalNames.push_back(definition->children[0]->string);
            terminalMap[definition->children[0]->string] = (unsigned int)(terminals.size() - 1);
        }
    }

    std::vector<ExtendedGrammar::Rule> rules;
    std::map<std::string, unsigned int> ruleMap;
    for(const auto &definition: node->children) {
        if(definition->type == DefNode::Type::Rule) {
            rules.push_back(ExtendedGrammar::Rule());
            ruleMap[definition->children[0]->string] = (unsigned int)(rules.size() - 1);
        }
    }

    for(const auto &definition: node->children) {
        if(definition->type == DefNode::Type::Rule) {
            std::string name = definition->children[0]->string;
            ExtendedGrammar::Rule &rule = rules[ruleMap[name]];
            rule.lhs = name;
            rule.rhs = createRhsNode(*definition->children[1], ruleMap, terminalMap, anonymousTerminalMap, terminals, terminalNames);
        }
    }

    Tokenizer::TokenValue endValue = (Tokenizer::TokenValue)terminals.size();
    terminals.push_back("");
    terminalNames.push_back("END");

    auto it = ruleMap.find("root");
    if(it == ruleMap.end()) {
        mParseError.message = "No <root> nonterminal defined";
        return;
    } else {
        ExtendedGrammar::Rule &rule = rules[it->second];
        std::unique_ptr<ExtendedGrammar::RhsNode> endNode = std::make_unique<ExtendedGrammar::RhsNodeSymbol>(ExtendedGrammar::RhsNodeSymbol::SymbolType::Terminal, endValue);
        if(rule.rhs->type == ExtendedGrammar::RhsNode::Type::Sequence) {
            static_cast<ExtendedGrammar::RhsNodeChildren&>(*rule.rhs).children.push_back(std::move(endNode));
        } else {
            std::unique_ptr<ExtendedGrammar::RhsNode> newNode = std::make_unique<ExtendedGrammar::RhsNodeChildren>(ExtendedGrammar::RhsNode::Type::Sequence, std::move(rule.rhs), std::move(endNode));
            rule.rhs = std::move(newNode);
        }

        Tokenizer::Configuration configuration;
        for(unsigned int i=0; i<terminals.size(); i++) {
            Tokenizer::Pattern pattern;
            pattern.regex = terminals[i];
            pattern.name = terminalNames[i];
            if(pattern.name == "IGNORE") {
                pattern.value = Tokenizer::InvalidTokenValue;
            } else {
                pattern.value = i;
            }
            configuration.patterns.push_back(std::move(pattern));
        }
        std::vector<Tokenizer::Configuration> configurations;
        configurations.push_back(std::move(configuration));
        mTokenizer = std::make_unique<Tokenizer>(std::move(configurations), endValue, Tokenizer::InvalidTokenValue);
        ExtendedGrammar extendedGrammar(std::move(terminalNames), std::move(rules), it->second);
        mGrammar = extendedGrammar.makeGrammar();
    }
}

bool DefReader::valid() const
{
    return mTokenizer && mGrammar;
}

const DefReader::ParseError &DefReader::parseError() const
{
    return mParseError;
}

const Tokenizer &DefReader::tokenizer() const
{
    return *mTokenizer;
}

const Grammar &DefReader::grammar() const
{
    return *mGrammar;
}

File: src/parser/def_reader.rs
use crate::parser;
use parser::ExtendedGrammar;
use parser::Tokenizer;

use std::io::BufRead;
use std::collections::HashMap;

pub struct DefReader {
    stream: parser::tokenizer::Stream,
    patterns: Vec<parser::tokenizer::Pattern>,
    rules: Vec<parser::extended_grammar::Rule>,
    anonymous_terminals: HashMap<String, usize>
}

pub struct ParseError {
    pub message: String,
    pub line: usize
}

impl ParseError {
    pub fn new(message: &str, line: usize) -> ParseError {
        ParseError { message: message.to_string(), line }
    }
}

#[derive(Copy, Clone)]
enum Token {
    Epsilon,
    Terminal,
    Nonterminal,
    Colon,
    Pipe,
    LParen,
    RParen,
    Star,
    Plus,
    Question,
    Literal,
    Regex,
    Newline,
    End
}

impl DefReader {
    pub fn parse(input: Box<dyn BufRead>) -> Result<(Tokenizer, ExtendedGrammar), ParseError> {
        let def_tokenizer = Self::make_def_tokenizer();

        let stream = parser::tokenizer::Stream::new(def_tokenizer, input);
        let mut def_reader = DefReader { stream, patterns: Vec::new(), rules: Vec::new(), anonymous_terminals: HashMap::new() };

        def_reader.parse_grammar()?;

        let mut terminals = Vec::new();
        for (i, pattern) in def_reader.patterns.iter_mut().enumerate() {
            if pattern.name == "IGNORE" {
                pattern.value = parser::tokenizer::INVALID_TOKEN_VALUE;
            } else {
                pattern.value = i;
            }

            terminals.push(pattern.name.clone());
        }
        let end_value = terminals.len();
        terminals.push(String::from("END"));

        let start_rule;
        if let Some(rule) = def_reader.rules.iter().position(|x| x.lhs == "root") {
            start_rule = rule;
        } else {
            return Err(ParseError::new("No <root> nonterminal defined", 0));
        }

        let configurations = vec![def_reader.patterns];
        let tokenizer = Tokenizer::new(configurations, end_value, parser::tokenizer::INVALID_TOKEN_VALUE);
        let extended_grammar = ExtendedGrammar::new(terminals, def_reader.rules, start_rule);
        return Ok((tokenizer, extended_grammar));
    }

    fn error_expected(&self, name: &str) -> ParseError {
        return ParseError::new(&format!("Expected: {name}"), self.stream.line);
    }

    fn error_expected_token(&self, token: Token) -> ParseError {
        let name  = match token {
            Token::Newline => "<newline>",
            _ => self.stream.pattern_name(token as usize)
        };

        return self.error_expected(name);
    }

    fn match_token_with_configuration(&mut self, token: Token, new_configuration: usize) -> Option<String> {
        let result;
        if self.stream.next_token().value == token as usize {
            result = Some(self.stream.next_token().text.clone());
            self.stream.set_configuration(new_configuration);
            self.stream.consume_token();
        } else {
            result = None;
        }
        return result;
    }

    fn match_token(&mut self, token: Token) -> Option<String> {
        return self.match_token_with_configuration(token, self.stream.configuration);
    }

    fn expect_token_with_configuration(&mut self, token: Token, new_configuration: usize) -> Result<String, ParseError> {
        if let Some(text) = self.match_token_with_configuration(token, new_configuration) {
            return Ok(text)
        } else {
            return Err(self.error_expected_token(token));
        }
    }

    fn expect_token(&mut self, token: Token) -> Result<String, ParseError> {
        return self.expect_token_with_configuration(token, self.stream.configuration);
    }

    fn parse_grammar(&mut self) -> Result<(), ParseError> {
        while !self.stream.is_end() {
            if let Some(pattern) = self.try_parse_pattern()? {
                if let Some(index) = self.patterns.iter().position(|x| x.name == pattern.name) {
                    self.patterns[index] = pattern;
                } else {
                    self.patterns.push(pattern);
                }
            } else if let Some(rule) = self.try_parse_rule()? {
                if let Some(index) = self.rules.iter().position(|x| x.lhs == rule.lhs) {
                    self.rules[index] = rule;
                } else {
                    self.rules.push(rule);
                }
            } else if let Some(_) = self.match_token(Token::Newline) {
                continue;
            } else {
                return Err(self.error_expected("item"));
            }
        }

        return Ok(());
    }

    fn try_parse_pattern(&mut self) -> Result<Option<parser::tokenizer::Pattern>, ParseError> {
        if let Some(text) = self.match_token(Token::Terminal) {
            self.expect_token_with_configuration(Token::Colon, 1)?;
            let regex = self.expect_token(Token::Regex)?;
            self.expect_token_with_configuration(Token::Newline, 0)?;
            return Ok(Some(parser::tokenizer::Pattern::new(&text, &regex, 0)));
        } else {
            return Ok(None);
        }
    }

    fn try_parse_rule(&mut self) -> Result<Option<parser::extended_grammar::Rule>, ParseError> {
        if let Some(text) = self.match_token(Token::Nonterminal) {
            self.expect_token(Token::Colon)?;
            let rhs = self.parse_rhs_options()?;
            self.expect_token(Token::Newline)?;
            let lhs = text[1..text.len() - 1].to_string();
            let rule = parser::extended_grammar::Rule { lhs, rhs };
            return Ok(Some(rule));
        } else {
            return Ok(None);
        }
    }

    fn parse_rhs_options(&mut self) -> Result<parser::extended_grammar::RHSNode, ParseError> {
        let mut options = Vec::new();
        loop { 
            options.push(self.parse_rhs_sequence()?);
            if let Some(_) = self.match_token(Token::Pipe) {
                continue;
            } else {
                break;
            }
        }
        let rhs_node;
        if options.len() == 1 {
            rhs_node = options.pop().unwrap();
        } else {
            rhs_node = parser::extended_grammar::RHSNode::OneOf(options);
        }

        return Ok(rhs_node);
    }

    fn parse_rhs_sequence(&mut self) -> Result<parser::extended_grammar::RHSNode, ParseError> {
        let mut items = Vec::new();

        while let Some(item) = self.try_parse_rhs_item()? {
            items.push(item);
        }

        if items.len() == 1 {
            return Ok(items.pop().unwrap());
        } else {
            return Ok(parser::extended_grammar::RHSNode::Sequence(items));
        }
    }

    pub fn try_parse_rhs_item(&mut self) -> Result<Option<parser::extended_grammar::RHSNode>, ParseError> {
        if let Some(symbol) = self.try_parse_rhs_symbol()? {
            let mut rhs_item = symbol;
            loop {
                if let Some(_) = self.match_token(Token::Star) {
                    rhs_item = parser::extended_grammar::RHSNode::ZeroOrMore(Box::new(rhs_item));
                } else if let Some(_) = self.match_token(Token::Plus) {
                    rhs_item = parser::extended_grammar::RHSNode::OneOrMore(Box::new(rhs_item));
                } else if let Some(_) = self.match_token(Token::Question) {
                    rhs_item = parser::extended_grammar::RHSNode::ZeroOrOne(Box::new(rhs_item));
                } else {
                    break;
                }
            }
            return Ok(Some(rhs_item));
        } else {
            return Ok(None);
        }
    }

    fn try_parse_rhs_symbol(&mut self) -> Result<Option<parser::extended_grammar::RHSNode>, ParseError> {
        if let Some(text) = self.match_token(Token::Terminal) {
            let pattern_name = text;
            let index;
            if let Some(i) = self.patterns.iter().position(|x| x.name == pattern_name) {
                index = i;
            } else {
                index = self.patterns.len();
                let dummy_pattern = parser::tokenizer::Pattern::new(&pattern_name, "", 0);
                self.patterns.push(dummy_pattern);
            }
            let symbol = parser::extended_grammar::Symbol::Terminal(index);
            return Ok(Some(parser::extended_grammar::RHSNode::Symbol(symbol)));
        } else if let Some(text) = self.match_token(Token::Nonterminal) {
            let rule_lhs = &text[1..text.len() - 1];
            let index;
            if let Some(i) = self.rules.iter().position(|x| x.lhs == rule_lhs) {
                index = i;
            } else {
                index = self.rules.len();
                let dummy_symbol = parser::extended_grammar::Symbol::Terminal(0);
                let dummy_rule = parser::extended_grammar::Rule { lhs: rule_lhs.to_string(), rhs: parser::extended_grammar::RHSNode::Symbol(dummy_symbol) };
                self.rules.push(dummy_rule);
            }
            let symbol = parser::extended_grammar::Symbol::Nonterminal(index);
            return Ok(Some(parser::extended_grammar::RHSNode::Symbol(symbol)));
        } else if let Some(text) = self.match_token(Token::Literal) {
            let literal_text = &text[1..text.len() - 1];

            let index;
            if self.anonymous_terminals.contains_key(literal_text) {
                index = self.anonymous_terminals[literal_text];
            } else {
                index = self.patterns.len();
                let pattern_name = format!("'{literal_text}'");
                let escaped_text = self.escape(&literal_text);
                let pattern = parser::tokenizer::Pattern::new(&pattern_name, &escaped_text, 0);
                self.patterns.push(pattern);
                self.anonymous_terminals.insert(literal_text.to_string(), index);
            }

            let symbol = parser::extended_grammar::Symbol::Terminal(index);
            return Ok(Some(parser::extended_grammar::RHSNode::Symbol(symbol)));
        } else if let Some(_) = self.match_token(Token::LParen) {
            let rhs_list = self.parse_rhs_options()?;
            self.expect_token(Token::RParen)?;
            return Ok(Some(rhs_list));
        } else {
            return Ok(None);
        }
    }

    fn make_def_tokenizer() -> Tokenizer {
        let pattern = |regex: &str, name: &str, value: Token| {
            parser::tokenizer::Pattern { regex: regex.to_string(), name: name.to_string(), value: value as usize}
        };

        let pattern_ignore = |regex: &str, name: &str| {
            parser::tokenizer::Pattern { regex: regex.to_string(), name: name.to_string(), value: parser::tokenizer::INVALID_TOKEN_VALUE}
        };

        let configurations = vec![
            vec![
                pattern("0", "<epsilon>", Token::Epsilon),
                pattern("\\w+", "<terminal>", Token::Terminal),
                pattern("<\\w+>", "<nonterminal>", Token::Nonterminal),
                pattern(":", ":", Token::Colon),
                pattern("\\|", "|", Token::Pipe),
                pattern("\\(", "(", Token::LParen),
                pattern("\\)", ")", Token::RParen),
                pattern("\\+", "+", Token::Plus),
                pattern("\\*", "*", Token::Star),
                pattern("\\?", "?", Token::Question),
                pattern("'[^']+'", "<literal>", Token::Literal),
                pattern_ignore("\\s", "<whitespace>")  
            ],
            vec![
                pattern("\\S+", "<regex>", Token::Regex),
                pattern_ignore("\\s", "<whitespace>")
            ]
        ];

        return Tokenizer::new(configurations, Token::End as usize, Token::Newline as usize);
    }

    fn escape(&self, input: &str) -> String {
        let mut result = String::new();
        for c in input.chars() {
            match c {
                ' ' => result += "\\s",
                '+' | '*' | '?' | '(' | ')' | '[' | ']' => { result.push('\\'); result.push(c); },
                _ => result.push(c)
            }
        }

        return result;
    }
}